packages/core/src/blitz-app-root.tsx

```diff
@@ -51,16 +51,19 @@ export function withBlitzInnerWrapper(Page: BlitzPage) {
     useAuthorizeIf(Page.authenticate === true)
 
     if (typeof window !== "undefined") {
-      // We read directly from publicDataStore.getData().userId instead of useSession
+      const publicData = publicDataStore.getData()
+      // We read directly from publicData.userId instead of useSession
       // so we can access userId on first render. useSession is always empty on first render
-      if (publicDataStore.getData().userId) {
+      if (publicData.userId) {
         clientDebug("[BlitzInnerRoot] logged in")
         let {redirectAuthenticatedTo} = Page
+        if (typeof redirectAuthenticatedTo === "function") {
+          redirectAuthenticatedTo = redirectAuthenticatedTo({session: publicData})
+        }
+        if (redirectAuthenticatedTo && typeof redirectAuthenticatedTo !== "string") {
```

Comment:
Should this be an `else if`?

Reply:
dont't think so. the second if block would never fire(in case if it's "else if") if redirectAuthenticatedTo is a function. but one of the possible return values from this function is an object, which one we should transform to a string inside the second if block on line 63. correct me if i wrong.

Reply:
@s-r-x I think @flybayer is correct because `redirectAuthenticatedTo` will be truthy since it's a function and `typeof redirectAuthenticatedTo !== "string"` will also be `true` because `"function" !== "string"`. 

Probably adding a test case would surface this? 

```ts
if (typeof redirectAuthenticatedTo === "function") {
  redirectAuthenticatedTo = redirectAuthenticatedTo({session: publicData})
  
}

// `redirectAuthenticatedTo` will be truthy since it's a function
//  and "function" !== "string" will equal true, so the second `if` block will run. 

if (redirectAuthenticatedTo && typeof redirectAuthenticatedTo !== "string") {
  redirectAuthenticatedTo = formatWithValidation(redirectAuthenticatedTo)
}
```

Reply:
Technically the code will work fine as is, but it's better to be an `else if` from a code readability standpoint. Easier to instantly see that only one of these branches will run.

Reply:
@erikshestopal,
@flybayer

https://user-images.githubusercontent.com/41614937/129136186-67c72eaf-30ca-4804-b706-a35227391d53.mp4

this is still fine because RedirectError accepts not only a string, but also RouteUrlObject(as far as i understand. at least in video it looks like it does) .
but the second if block is never fire in this case, and tbh i'm not quite sure how important that formatWithValidation call is.

Reply:
@flybayer i simplified the code, make everything immutable